File: LeetCode/iter1/c++/589.n-ary-tree-preorder-traversal.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Node {
public:
	int val;
	vector<Node*> children;

	Node() {}

	Node(int _val, vector<Node*> _children) {
		val = _val;
		children = _children;
	}
};

class Solution {
public:
	//iterative
	vector<int> preorder(Node* root) {
		vector<int> ret;

		vector<Node*> nodeV;
		nodeV.push_back(root);

		while (!nodeV.empty()) {
			Node* node = nodeV.back();
			nodeV.pop_back();

			if (node != NULL) {
				ret.push_back(node->val);
				for (int i = node->children.size() - 1; i >= 0; i--) {
					nodeV.push_back(node->children[i]);
				}
			}
		}

		return ret;
	}

	//recursion
	//vector<int> preorder(Node* root) {
	//	vector<int> ret;
	//	if (root == NULL) {
	//		return ret;
	//	}

	//	ret.push_back(root->val);
	//	for (auto childNode : root->children) {
	//		vector<int> childPreOrder = preorder(childNode);
	//		for (int tmpVal : childPreOrder) {
	//			ret.push_back(tmpVal);
	//		}
	//	}

	//	return ret;
	//}
};

int main()
{
	vector<Node*> v5;
	Node *node5 = new Node(5, v5);

	vector<Node*> v6;
	Node *node6 = new Node(6, v6);

	vector<Node*> v3;
	v3.push_back(node5);
	v3.push_back(node6);
	Node *node3 = new Node(3, v3);

	vector<Node*> v2;
	Node *node2 = new Node(2, v2);

	vector<Node*> v4;
	Node *node4 = new Node(4, v4);

	vector<Node*> v1;
	v1.push_back(node3);
	v1.push_back(node2);
	v1.push_back(node4);
	Node *node1 = new Node(1, v1);

	Solution solution;
	vector<int> ret = solution.preorder(node1);
	for (int num : ret) {
		cout << num << " ";
	}

	return 0;
}

File: LeetCode/iter1/java/589.n-ary-tree-preorder-traversal.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Solution {

    public static void main(String[] args) {
        Node node5 = new Node(5, null);
        Node node6 = new Node(6, null);
        List<Node> lst3 = new ArrayList<>();
        lst3.add(node5);
        lst3.add(node6);
        Node node3 = new Node(3, lst3);

        Node node2 = new Node(2, null);
        Node node4 = new Node(4, null);
        List<Node> lst1 = new ArrayList<>();
        lst1.add(node3);
        lst1.add(node2);
        lst1.add(node4);

        Node node1 = new Node(1, lst1);

        System.out.println(preorder(node1));
    }

    public static List<Integer> preorder(Node root) {
        if (root == null) {
            return new ArrayList<>();
        }

        List<Integer> ret = new ArrayList<>();
        Stack<Node> stack = new Stack<>();
        stack.add(root);

        while (!stack.isEmpty()) {
            Node tmp = stack.pop();
            if (tmp == null) {
                continue;
            }
            ret.add(tmp.val);

            if (tmp.children == null) {
                continue;
            }
            for (int i = tmp.children.size() - 1; i >= 0; i--) {
                stack.push(tmp.children.get(i));
            }
        }

        return ret;
    }

    //recursion
//    public static List<Integer> preorder(Node root) {
//        if (root == null) {
//            return new ArrayList<>();
//        }
//
//        List<Integer> ret = new ArrayList<>();
//        ret.add(root.val);
//        if (root.children == null) {
//            return ret;
//        }
//        for (Node node : root.children) {
//            List<Integer> tmpLst = preorder(node);
//            for (int num : tmpLst) {
//                ret.add(num);
//            }
//        }
//
//        return ret;
//    }

    static class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }
}